src/react/apiBase.ts

```diff
@@ -1,51 +1,36 @@
 import {Store, is, Scope} from 'effector'
 import React from 'react'
-import {useIsomorphicLayoutEffect} from './useIsomorphicLayoutEffect'
+import {useSyncExternalStore} from 'use-sync-external-store/shim'
+import {useSyncExternalStoreWithSelector} from 'use-sync-external-store/shim/with-selector'
 import {throwError} from './throw'
 import {createWatch} from './createWatch'
 import {withDisplayName} from './withDisplayName'
 
 const stateReader = <T>(store: Store<T>, scope?: Scope) =>
   scope ? scope.getState(store) : store.getState()
 const basicUpdateFilter = <T>(upd: T, oldValue: T) => upd !== oldValue
-const createNotifier = () =>
-  React.useReducer((n: any, action: void) => n + 1, 0)[1]
+const keysEqual = (a?: readonly any[], b?: readonly any[]) => {
+  if (!a || !b || a.length !== b.length) return false
+
+  let isEqual = true
+
+  for (let i = 0; i < a.length; i++) {
+    if (a[i] !== b[i]) {
+      isEqual = false
+      break
+    }
+  }
+
+  return isEqual
+}
 
 export function useStoreBase<State>(store: Store<State>, scope?: Scope) {
   if (!is.store(store)) throwError('expect useStore argument to be a store')
 
-  const currentValue = stateReader(store, scope)
-  const inc = createNotifier()
-  const currentStore = React.useRef({
-    store,
-    value: currentValue,
-    pending: false,
-  })
-  useIsomorphicLayoutEffect(() => {
-    const stop = createWatch(
-      store,
-      upd => {
-        const ref = currentStore.current
-        if (!ref.pending) {
-          ref.value = upd
-          ref.pending = true
-          inc()
-          ref.pending = false
-        }
-      },
-      scope,
-    )
-    const newValue = stateReader(store, scope)
-    const ref = currentStore.current
-    if (ref.store === store && ref.value !== newValue) {
-      ref.value = newValue
-      ref.pending = true
-      inc()
-      ref.pending = false
-    }
-    ref.store = store
-    return stop
-  }, [store, scope])
+  const subscribe = (cb: () => void) => createWatch(store, cb, scope)
+  const read = () => stateReader(store, scope)
+  const currentValue = useSyncExternalStore(subscribe, read, read)
```

Comment:
Third argument is `getServerSnapshot` - in case of effector there is no difference with client version
https://reactjs.org/docs/hooks-reference.html#usesyncexternalstore